lightning/src/util/ser.rs

```diff
@@ -166,18 +172,24 @@ impl<R: Read> Read for ReadTrackingReader<R> {
 /// (C-not exported) as we only export serialization to/from byte arrays instead
 pub trait Writeable {
 	/// Writes self out to the given Writer
-	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error>;
+	fn write(&self, writer: &mut Writer) -> Result<(), ::std::io::Error>;
 
 	/// Writes self out to a Vec<u8>
 	fn encode(&self) -> Vec<u8> {
-		let mut msg = VecWriter(Vec::new());
+		let mut msg = Writer::new();
 		self.write(&mut msg).unwrap();
 		msg.0
 	}
 
+	/// Calculates the encoded length of this object.
+	/// This may be overridden to optimize for specific structs.
```

Comment:
Hmmmm, this feels like having two APIs for the same thing - size_hint and encoding_length. size_hint is a "push" and this is a "pull", but given we only ever have one `Writer` interface, I'd think size_hint is the way to go. I'd rather just expose this directly on OnionHopData and maybe add a new (non-public) trait for the test harness stuff.

Reply:
done